Completion of code challenge

cdevroe opened the pull request:
This was a fun and interesting challenge. Most of the challenge, for someone like me that is new to Gutenberg block creation, is in the fact that most of the documentation around block development is out of date. But, I don't know what I don't know, so perhaps this PR will show that I totally missed the point of the exercise.

Some things I'm not sure about:
- In the screenshot the title seems to be either selected or highlighted. I wasn't certain if that was to be styled that way or not. Based on the appearance of the highlight not covering each of the letters on each row, my guess is that this should somehow be added. I am unsure how I would add this.
- I decided to strip the HTML from the excerpt so that the display would remain consistent if there were several related posts. I'm unsure if this was the right call or if I should have allowed some basic HTML rendering even though the excerpt was truncated.

A few things I omitted that I would likely do:
- As the readme references, I'd add a client-side lookup to refresh the related posts list when the website was viewed. I'd cache the results so that the lookup wouldn't occur on all page views.
- Some of the dev dependencies need to be updated (the repo being 8 months old makes this make sense)

I appreciated the opportunity to take part in this challenge.

cdevroe committed:
Linting, added useBlockProps, and fixed some spelling and gramatical errors.

The code in edit.js and preview.js was not properly formatted, sometimes in the incorrect order, and had some spelling mistakes in the view render. I ran ES Lint on both edit and preview, added useBlockProps to the edit view so that the block was properly wrapped - which turned the toolbar back on. I also renamed the icon variables to be more descriptive to potentially cut down on any confusion in the future.

cdevroe committed:
Added apiFetch, changed how postId is determined

Related posts were being fetched using the window.fetch method rather than WordP
ress' apiFetch. I added @wordpress/api-fetch as a dev dependency, rewrote the ca
ll to the REST API to use that. Also, the block was trying to save the incorrect response for the posts. The postId was being determined using the
URL and regex. I'm not certain, but this seems like it could cause an issue. So
instead I am pulling it in via the context attribute.

cdevroe committed:
Updated excerpt preview

Added a utility function to format the excerpt for both length and to remove any HTML from it. The excerpt can be truncated to any desired length.

cdevroe committed:
Updated countryCode selector, added ABBR

The country code in the screenshot appears to be wrapped in an abbreviation HTML tag - perhaps for accessibility reasons? To allow for that the selector for saving the countryCode needed to be updated. I decided to use a data attribute. I also added styles for the new tag.

cdevroe committed:
Class name and style adjustments

The class names for the related posts were not consistent with the rest of the custom block. The class names were adjusted as well as the selectors for the block attributes.

cdevroe committed:
Additional style changes and a bug fix

The SASS file's selectors weren't nested properly. I also missed a selector which thankfully didn't effect the render of the block. This is fixed.

bartoszgadomski commented on src/edit.js:
```diff
@@ -35,48 +58,64 @@ export default function Edit( { attributes, setAttributes } ) {
-				throw new Error( `HTTP error! Status: ${ response.status }` );
-
-			const posts = await response.json();
+			const posts = await apiFetch( {
+				path: `/wp/v2/posts?search=${ countries[ countryCode ] }&exclude=${ postId }`,
+			} ).then( ( response ) => {
+				return response;
+			} );
```
@cdevroe Three questions here:

1. What are the benefits of using `apiFetch` compared to `window.fetch`?
2. Is there a more robust way to fetch data than `apiFetch`?
3. How could the related posts fetching be optimized even more (in terms of payload size)?

cdevroe replied:
> 1. What are the benefits of using `apiFetch` compared to `window.fetch`?

When I was looking into how to fix the REST API call to fetch the related posts I saw apiFetch was officially supported, regularly updated, etc. It seems to have some nice features specific to WordPress such as dealing with nonce when needed.

> 2. Is there a more robust way to fetch data than `apiFetch`?

There are other third-party clients that could be used to fetch data. apiFetch allows for custom fetch handlers that have better error handling, security features built-in, etc. I didn’t think it was warranted to bring in an entirely different fetch handler since we are calling from the local WordPress REST API. If I were pulling data from an external API, especially one that could potentially return security problems, I’d use something more robust.

> 3. How could the related posts fetching be optimized even more (in terms of payload size)?

I see two ways to reduce the size of the payload. One is to return only posts (on the off chance other types would be returned) and also using the `_fields` argument to limit the number of fields returned to only that which we will use. I’ve added this to the PR.

bartoszgadomski commented on src/block.json:
```diff
@@ -13,39 +13,41 @@
 			}
 		}
 	},
+	"usesContext": [ "postId" ],
```
@cdevroe Can you tell more about block context? How does block context work internally? Is the context available only in React or in PHP, too?

cdevroe replied:
As I understand it, Block context works similarly to how Reacts global context works. If you need nested components to retrieve some State, rather than needing to propagate all the way through the tree you can use context. I would likely use this feature sparingly. But I thought using context to determine the currently loaded postId was a good use of this feature, rather than relying on the URL to have the postId.

I didn’t prove this out in code, but I believe if we were using a render callback function in PHP to display the block on the front end we could access the postId using the block context.

bartoszgadomski commented on src/utils.js:
```diff
@@ -6,3 +6,30 @@ export function getEmojiFlag( countryCode ) {
+		Else 
+			Empty string
+*/
+export function formatExcerpt( excerpt, excerptLength ) {
+	if ( ! excerpt ) return '';
+
+	if ( excerpt.length > excerptLength ) {
+		return (
+			excerpt.replace( /<[^>]*>?/gm, '' ).substring( 0, excerptLength ) +
+			'...'
+		);
+	}
+	return '';
+}
```
@cdevroe What are the other ways to sanitize HTML markup in JavaScript? Assume that the client wants to preserve HTML tags in the output.

cdevroe replied:
If there are built in WordPress JavaScript utility functions to do something like this I couldn't find one. One way could be to create a list of accepted HTML tags that would be allowed in the Excerpt. Such as EM, STRONG, etc. And after I’ve truncated the Excerpt to the appropriate size, close any unclosed HTML tags found in the truncated excerpt. There is a simple, yet perhaps hacky?, way to do this by creating a function that uses `.innerHTML` to do so.

bartoszgadomski commented on src/save.js:
```diff
@@ -1,3 +1,6 @@
+ */
 import Preview from './preview';
 
 export default function Save( { attributes } ) {
```
@cdevroe Two questions here:

1. How can we pass any extra custom HTML attributes (like class names) to the block element wrapper?
2. How to ensure that the block class name is rendered properly?

cdevroe replied:
> 1. How can we pass any extra custom HTML attributes (like class names) to the block element wrapper?

Good catch. Sorry. Pulling in the block properties will take any custom classes added in the editor and apply them to the wrapping DIV. I’ve added the code for this to the PR.

> 2. How to ensure that the block class name is rendered properly?

I’m sorry, I don't fully understand this question. 🤔

bartoszgadomski commented on src/preview.js:
```diff
@@ -1,51 +1,76 @@
-				{ ' ' }
-				(<span className="xwp-country-card__country-code">{ countryCode }</span>),
-				{ ' ' }
-				{ continentNames[continents[countryCode]] }!
+				{ __( 'Hello from' ) }{ ' ' }
+				<strong>{ countries[ countryCode ] }</strong> (
+				<span
+					className="xwp-country-card__country-code"
+					data-country-code={ countryCode }
+				>
+					<abbr title={ countries[ countryCode ] }>
+						{ countryCode }
+					</abbr>
+				</span>
+				), { continentNames[ continents[ countryCode ] ] }!
 			</h3>
```
@cdevroe Do you see potential i18n issues with the card heading? What prevented you from making the string easier to localize? Do you have an idea of how this should be handled?

cdevroe replied:
Yes I do! And I totally missed that these strings didn't include the text-domain. So I've added those to the PR.

When I saw this code I believe my assumption was that the space between “Hello from” and the country name was to be kept separate from the text definition? But I’m unsure on that. So I do not know the proper way to handle trailing spaces in localized strings.

bartoszgadomski commented on src/edit.js:
```diff
@@ -1,19 +1,42 @@
-        value: code,
-        label:  getEmojiFlag( code ) + '  ' + countries[code] + ' — ' + code
-	}));
+	const options = Object.keys( countries ).map( ( code ) => ( {
+		value: code,
+		label: getEmojiFlag( code ) + '  ' + countries[ code ] + ' — ' + code,
+	} ) );
```
@cdevroe Do you think the options array calculation could be improved in some way?

cdevroe replied:
As a matter of style I suppose different ways to iterate over the countries object could be looked into such as `for…in` or using `Object.entries`, etc. JavaScript seems to have many ways to do this. I wouldn't be confident in saying that one would be faster than another in this context since the dataset is rather small.

However, I think the biggest improvement would be adding a `.sort` method to list the options for the user in alphabetical order by key. I’ve included that code in this PR.

bartoszgadomski reviewed (commented):
@cdevroe Thank you for completing the challenge! I added a few questions as inline comments.

Was there something in the challenge code that you didn’t fix even though you noticed a potential issue (e.g. due to lack of time or instructions/expectations not being clear enough)?

CC: @emilykaija @delawski

cdevroe committed:
Sorting country list options

Added sorting for the list of countries available to the user. Listed now in alphabetical order by Country Code.

cdevroe committed:
Limiting API response size

The related posts API call has been slimmed down to only include the fields we're using in the plugin.

cdevroe committed:
Adding wrapper classes to saved output

The saved block didn't have the appropriate wrapper classes. Imported useBlockProps and added those to the wrapping DIV. This allows for custom classes to be added to the block via the editor.

cdevroe committed:
Added text domain to localized strings in Preview

The localized strings in the Preview were missing the text-domain.
